packages/trie/src/proof/range.ts

```diff
@@ -382,6 +384,180 @@ async function hasRightElement(trie: Trie, key: Nibbles): Promise<boolean> {
   return false
 }
 
+export function isValueNode(node: TrieNode): boolean {
+  return (
+    node instanceof LeafNode ||
+    (node instanceof BranchNode && node.value() !== null && node.value()!.length > 0)
+  )
+}
+
+export function hasRightBranch(
+  node: BranchNode,
+  nibble: number
+): undefined | [number, EmbeddedNode] {
+  for (let i = nibble + 1; i < 16; i++) {
+    try {
+      const b = node.getBranch(i)
+      if (b !== null) {
+        return [i, b]
+      }
+    } catch (err) {
+      //
+    }
+  }
+  return undefined
+}
+
+export async function walkRight(this: Trie, key: Nibbles, hash: Uint8Array, childIdx: number) {
+  const rightWalk = this.walkTrieIterable(hash, [childIdx])
+  for await (const { node, currentKey } of rightWalk) {
+    if (nibblesCompare(key, [...currentKey, ...('_nibbles' in node ? node._nibbles : [])]) < 0) {
+      if (node instanceof LeafNode && node.value()!.length < 0) {
+        const cur = [...currentKey, ...('_nibbles' in node ? node._nibbles : [])]
+        return {
+          node,
+          currentKey: cur,
+        }
+      }
+    }
+  }
+  return null
+}
+
+export async function rightNode(
+  this: Trie,
+  key: Nibbles,
+  path: Path
+): Promise<{ node: TrieNode; currentKey: Nibbles } | null> {
```

Comment:
If the keys are all equal size, then yes, because all value nodes will terminate.

If keys were not the same size, then we would end up finding value nodes that did not terminate, and end up with leftover nibbles.